theme/utils.py

```diff
@@ -29,56 +35,146 @@ def get_upload_path_userprofile(instance, filename):
-            # return quota grace period message
-            cut_off_date = date.today() + timedelta(days=uq.remaining_grace_period)
-            msg_template_str = '{}{}\n'.format(qmsg.grace_period_content_prepend, qmsg.content)
-            return_msg += msg_template_str.format(used=rounded_used_val,
-                                                  unit=uq.unit,
-                                                  allocated=uq.allocated_value,
-                                                  zone=uq.zone,
-                                                  percent=rounded_percent,
-                                                  cut_off_date=cut_off_date)
-        elif percent >= soft_limit:
-            # return quota warning message
-            msg_template_str = '{}{}\n'.format(qmsg.warning_content_prepend, qmsg.content)
-            return_msg += msg_template_str.format(used=rounded_used_val,
-                                                  unit=uq.unit,
-                                                  allocated=uq.allocated_value,
-                                                  zone=uq.zone,
-                                                  percent=rounded_percent)
-        else:
-            # return quota informational message
-            return_msg += qmsg.warning_content_prepend.format(allocated=uq.allocated_value,
-                                                              unit=uq.unit,
-                                                              used=rounded_used_val,
-                                                              zone=uq.zone,
-                                                              percent=rounded_percent)
-        return return_msg
+    today = date.today()
+    grace = uq.grace_period_ends
+    allocated = uq.allocated_value
+    unit = uq.unit
+    uz, dz = uq.get_used_value_by_zone()
+    used = uz + dz
+    uzp = uz * 100.0 / allocated
+    dzp = dz * 100.0 / allocated
+    percent = used * 100.0 / allocated
+    remaining = allocated - used
+
+    # initiate grace_period counting if not already started by the daily celery task
+    if percent >= 100 and not grace:
+        # This would indicate that the grace period has not been set even though the user went over quota.
+        # This should not happen.
+        logger.error(f"User {uq.user.username} went over quota but grace period was not set.")
+        status = QuotaStatus.GRACE_PERIOD
+
+    if percent >= hard_limit or (percent >= 100 and grace <= today):
+        status = QuotaStatus.ENFORCEMENT
+    elif percent >= 100 and grace > today:
+        status = QuotaStatus.GRACE_PERIOD
+    elif percent >= soft_limit:
+        status = QuotaStatus.WARNING
+    else:
+        status = QuotaStatus.INFO
+
+    uq_data = {"used": used,
+               "allocated": allocated,
+               "unit": unit,
+               "uz": uz,
+               "dz": dz,
+               "uz_percent": uzp if uzp < 100 else 100,
```

Comment:
I wanted to limit to 100% and then use `percent_over` to see how much over. This was because of how display is implemented on the front end. But you're right that we could just use `dz_percent` and `uz_percent` and allow them to go over 100, and remove `percent_over` (and modify the frontend to suite). Do you think that's preferable?